README.md
# Modern URI

[![License BSD3](https://img.shields.io/badge/license-BSD3-brightgreen.svg)](http://opensource.org/licenses/BSD-3-Clause)
[![Hackage](https://img.shields.io/hackage/v/modern-uri.svg?style=flat)](https://hackage.haskell.org/package/modern-uri)
[![Stackage Nightly](http://stackage.org/package/modern-uri/badge/nightly)](http://stackage.org/nightly/package/modern-uri)
[![Stackage LTS](http://stackage.org/package/modern-uri/badge/lts)](http://stackage.org/lts/package/modern-uri)
![CI](https://github.com/mrkkrp/modern-uri/workflows/CI/badge.svg?branch=master)

This is a modern library for working with URIs in Haskell as per RFC 3986:

https://tools.ietf.org/html/rfc3986

## Features

The `modern-uri` package features:

* Correct by construction `URI` data type. The correctness is ensured by
  making sure that every sub-component of the `URI` record is by itself
  cannot be invalid. This boils down to careful use of types and a set of
  smart constructors.
* Textual components in the `URI` data type are represented as `Text` rather
  than `ByteString`, because they are percent-decoded and so they can
  contain characters outside of ASCII range (i.e. Unicode). This allows for
  easier manipulation of `URI`s, while encoding and decoding headaches are
  handled by the parsers and renders for you.
* Absolute and relative URIs differ only by the scheme component: if it's
  `Nothing`, then URI is relative, otherwise it's absolute.
* Megaparsec parser that can be used as a standalone smart constructor for
  the `URI` data type (see `mkURI`) as well as be seamlessly integrated into
  a bigger Megaparsec parser that consumes strict `Text` (see `parser`) or
  strict `ByteString` (see `parserBs`).
* The parser performs some normalization, for example it collapses
  consecutive slashes. Some smart constructors such as `mkScheme` and
  `mkHost` also perform normalization. So in a sense URIs are also
  “normalized by construction” to some extent.
* Fast rendering to strict `Text` and `ByteString` as well as to their
  respective `Builder` types and to `String`/`ShowS`.
* Extensive set of lensy helpers for easier manipulation of the nested data
  types (see `Text.URI.Lens`).
* Quasi-quoters for compile-time construction of the `URI` data type and
  refined text types (see `Text.URI.QQ`).

## Quick start

The `modern-uri` package serves three main purposes:

* Construction of the `URI` data type.
* Inspection and manipulation of the `URI` data type (in the sense of
  changing its parts).
* Rendering of `URI`s.

Let's walk through every operation quickly.

### Construction of `URI`s

There are four ways to create a `URI` value. First off, one could assemble
it manually like so:

```haskell
λ> :set -XOverloadedStrings
λ> import qualified Text.URI as URI
λ> scheme <- URI.mkScheme "https"
λ> scheme
"https"
λ> host <- URI.mkHost "markkarpov.com"
λ> host
"markkarpov.com"
λ> let uri = URI.URI (Just scheme) (Right (URI.Authority Nothing host Nothing)) Nothing [] Nothing
λ> uri
URI
  { uriScheme = Just "https",
    uriAuthority =
      Right
        ( Authority
            { authUserInfo = Nothing,
              authHost = "markkarpov.com",
              authPort = Nothing
            }
        ),
    uriPath = Nothing,
    uriQuery = [],
    uriFragment = Nothing
  }
```

In this library we use quite a few refined text values. They only can be
constructed by using smart constructors like `mkScheme :: MonadThrow m =>
Text -> m (RText 'Scheme)`. For example, if argument to `mkScheme` is not a
valid scheme, an exception will be thrown. Note that monads such as `Maybe`
are also instances of the `MonadThrow` type class, and so the smart
constructors may be used in pure setting as well.

There is a smart constructor that can make an entire `URI` too, it's called
(unsurprisingly) `mkURI`:

```haskell
λ> uri <- URI.mkURI "https://markkarpov.com"
λ> uri
URI
  { uriScheme = Just "https",
    uriAuthority =
      Right
        ( Authority
            { authUserInfo = Nothing,
              authHost = "markkarpov.com",
              authPort = Nothing
            }
        ),
    uriPath = Nothing,
    uriQuery = [],
    uriFragment = Nothing
  }
```

If the argument of `mkURI` is not a valid URI, then an exception will be
thrown. The exception will contain full context and the actual parse error.

If some refined text value or `URI` is known statically at compile time, we
can use Template Haskell, namely the “quasi quotes” feature. To do so import
the `Text.URI.QQ` module and enable the `QuasiQuotes` language extension,
like so:

```haskell
λ> :set -XQuasiQuotes
λ> import qualified Text.URI.QQ as QQ
λ> let uri = [QQ.uri|https://markkarpov.com|]
λ> uri
URI
  { uriScheme = Just "https",
    uriAuthority =
      Right
        ( Authority
            { authUserInfo = Nothing,
              authHost = "markkarpov.com",
              authPort = Nothing
            }
        ),
    uriPath = Nothing,
    uriQuery = [],
    uriFragment = Nothing
  }
```

Note how the value returned by the `url` quasi quote is pure, its
construction cannot fail because when there is an invalid URI inside the
quote it's a compilation error. The `Text.URI.QQ` module has quasi-quoters
for scheme, host, and other components.

Finally, the package provides two Megaparsec parsers: `parser` and
`parserBs`. The first works on strict `Text`, while other one works on
strict `ByteString`s. You can use the parsers in a bigger Megaparsec parser
to parse `URI`s.

### Inspection and manipulation

Although one could use record syntax directly, possibly with language
extensions like `RecordWildcards`, the best way to inspect and edit parts of
`URI` is with lenses. The lenses can be found in the `Text.URI.Lens` module.
If you have never used the
[`lens`](https://hackage.haskell.org/package/lens) library, you could
probably start by reading/watching materials suggested in the library
description on Hackage.

Here are some examples, just to show off what you can do:

```haskell
λ> import Text.URI.Lens
λ> uri <- URI.mkURI "https://example.com/some/path?foo=bar&baz=quux&foo=foo"
λ> uri ^. uriScheme
Just "https"
λ> uri ^? uriAuthority . _Right . authHost
Just "example.com"
λ> uri ^. isPathAbsolute
True
λ> uri ^. uriPath
["some","path"]
λ> k <- URI.mkQueryKey "foo"
λ> uri ^.. uriQuery . queryParam k
["bar","foo"]
-- etc.
```

### Rendering

Rendering turns a `URI` into a sequence of bytes or characters. Currently
the following options are available:

* `render` for rendering to strict `Text`.
* `render'` for rendering to text `Builder`. It's possible to turn that into
  lazy `Text` by using the `toLazyText` function from
  `Data.Text.Lazy.Builder`.
* `renderBs` for rendering to strict `ByteString`.
* `renderBs'` for rendering to byte string `Builder`. Similarly it's
  possible to get a lazy `ByteString` from that by using the
  `toLazyByteString` function from `Data.ByteString.Builder`.
* `renderStr` can be used to render to `String`. Sometimes it's handy. The
  render uses difference lists internally so it's not that slow, but in
  general I'd advise avoiding `String`s.
* `renderStr'` returns `ShowS`, which is just a synonym for `String ->
  String`—a function that prepends the result of rendering to a given
  `String`. This is useful when the `URI` you want to render is a part of a
  bigger output, just like with the builders mentioned above.

Examples:

```haskell
λ> uri <- mkURI "https://markkarpov.com/posts.html"
λ> render uri
"https://markkarpov.com/posts.html"
λ> renderBs uri
"https://markkarpov.com/posts.html"
λ> renderStr uri
"https://markkarpov.com/posts.html"
-- etc.
```

## Contribution

Issues, bugs, and questions may be reported in [the GitHub issue tracker for
this project](https://github.com/mrkkrp/modern-uri/issues).

Pull requests are also welcome.

## License

Copyright © 2017–present Mark Karpov

Distributed under BSD 3 clause license.


CHANGELOG.md
## 0.3.6.1

* Host now can contain unreserved characters rather than simply alpha
  numeric ones, which was against RFC 3986. [Issue
  73](https://github.com/mrkkrp/modern-uri/issues/73).

## 0.3.6.0

* Now colons are not escaped in paths, unless the `URI` in question is a
  URI-reference, in which case colons in the first path segment are escaped.
  See [RFC 3986, section
  3.3](https://www.rfc-editor.org/rfc/rfc3986#section-3.3). [Issue
  55](https://github.com/mrkkrp/modern-uri/issues/55).

## Modern URI 0.3.5.0

* Added `Hashable` instances for `URI`, `Authority`, `UserInfo`,
  `QueryParam`, `RText`.

## Modern URI 0.3.4.4

* The `mailto` scheme does not escape `@` in its paths (fixes the regression
  introduced in 0.3.4.3).

## Modern URI 0.3.4.3

* Percent encode delimiter characters and `@` that appear in a path
  component. [PR 47](https://github.com/mrkkrp/modern-uri/pull/47).

* Sub-domains that look like IPv4 can now be parsed. [Issue
  46](https://github.com/mrkkrp/modern-uri/issues/46).

## Modern URI 0.3.4.2

* Improved handling of percent-encoded sequences of bytes that cannot be
  decoded as UTF-8 text. Now friendly error messages are reported in these
  cases.

## Modern URI 0.3.4.1

* Works with GHC 9.0.1.

## Modern URI 0.3.4.0

* URIs with authority component and without path are now rendered without
  trailing slashes.

## Modern URI 0.3.3.1

* Works with `bytestring-0.11`.

## Modern URI 0.3.3.0

* Added `mkURIBs` for parsing `ByteString` as a `URI`.

## Modern URI 0.3.2.0

* Quasi-quoters from `Text.URI.QQ` now can be used in pattern context when
  the `ViewPatterns` extension is enabled.

* Dropped support for GHC 8.2.x.

## Modern URI 0.3.1.0

* Dropped support for GHC 8.0 and 7.10.

* Added Template Haskell `Lift` instance for the `URI` type and its
  sub-components.

## Modern URI 0.3.0.1

* Allow superfluous `&` right after question sign in query parameters.

## Modern URI 0.3.0.0

* Uses Megaparsec 7. Visible API changes amount to an adjustment in
  definition of the `ParseException` type.

## Modern URI 0.2.2.0

* Removed a potentially overlapping instance `Arbitrary (NonEmpty (RText
  'PathPiece))`.

* Fixed a bug that made it impossible to have empty host names. This allows
  us to parse URIs like `file:///etc/hosts`.

## Modern URI 0.2.1.0

* Added `emptyURI`—`URI` value representing the empty URI.

## Modern URI 0.2.0.0

* Changed the type of `uriPath` field of the `URI` record from `[RText
  'PathPiece]` to `Maybe (Bool, NonEmpty (RText 'PathPiece))`. This allows
  us to store whether there is a trailing slash in the path or not. See the
  updated documentation for more information.

* Added the `relativeTo` function.

* Added the `uriTrailingSlash` 0-1 traversal in `Text.URI.Lens`.

## Modern URI 0.1.2.1

* Allow Megaparsec 6.4.0.

## Modern URI 0.1.2.0

* Fixed handling of `+` in query strings. Now `+` is parsed as space and
  serialized as `%2b` as per RFC 1866 (paragraph 8.2.1). White space in
  query parameters is serialized as `+`.

## Modern URI 0.1.1.1

* Fixed implementation of `Text.URI.Lens.queryParam` traversal.

## Modern URI 0.1.1.0

* Derived `NFData` for `ParseException`.

* Adjusted percent-encoding in renders so it's only used when absolutely
  necessary. Previously we percent-escaped a bit too much, which, strictly
  speaking, did not make the renders incorrect, but that didn't look nice
  either.

## Modern URI 0.1.0.1

* Updated the readme to include “Quick start” instructions and some
  examples.

## Modern URI 0.1.0.0

* Changed the type of `uriAuthority` from `Maybe Authority` to `Either Bool
  Authority`. This allows to know if URI path is absolute or not without
  duplication of information, i.e. when the `Authority` component is present
  the path is necessarily absolute, otherwise the `Bool` value tells if it's
  absolute (`True`) or relative (`False`).

* Added `isPathAbsolute` in `Text.URI` and the corresponding getter in
  `Text.URI.Lens`.

## Modern URI 0.0.2.0

* Added the `renderStr` and `renderStr'` functions for efficient rendering
  to `String` and `ShowS`.

* Added the `parserBs` that can consume strict `ByteString` streams.

## Modern URI 0.0.1.0

* Initial release.


Text/URI/Types.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskellQuotes #-}
{-# LANGUAGE TupleSections #-}

-- |
-- Module      :  Text.URI.Types
-- Copyright   :  © 2017–present Mark Karpov
-- License     :  BSD 3 clause
--
-- Maintainer  :  Mark Karpov <markkarpov92@gmail.com>
-- Stability   :  experimental
-- Portability :  portable
--
-- 'URI' types, an internal module.
module Text.URI.Types
  ( -- * Data types
    URI (..),
    makeAbsolute,
    isPathAbsolute,
    Authority (..),
    UserInfo (..),
    QueryParam (..),
    ParseException (..),
    ParseExceptionBs (..),

    -- * Refined text
    RText,
    RTextLabel (..),
    mkScheme,
    mkHost,
    mkUsername,
    mkPassword,
    mkPathPiece,
    mkQueryKey,
    mkQueryValue,
    mkFragment,
    unRText,
    RTextException (..),

    -- * Utils
    pHost,
  )
where

import Control.DeepSeq
import Control.Monad
import Control.Monad.Catch (Exception (..), MonadThrow (..))
import Data.ByteString (ByteString)
import Data.Char
import Data.Data (Data)
import Data.Either (fromLeft)
import Data.Hashable (Hashable)
import Data.List (intercalate)
import Data.List.NonEmpty (NonEmpty (..))
import qualified Data.List.NonEmpty as NE
import Data.Maybe (fromJust, fromMaybe, isJust)
import Data.Proxy
import Data.Text (Text)
import qualified Data.Text as T
import Data.Typeable (Typeable, cast)
import Data.Void
import Data.Word (Word16, Word8)
import GHC.Generics
import qualified Language.Haskell.TH.Syntax as TH
import Numeric (showHex, showInt)
import Test.QuickCheck
import Text.Megaparsec
import Text.URI.Parser.Text.Utils (pHost)

----------------------------------------------------------------------------
-- Data types

-- | Uniform resource identifier (URI) reference. We use refined 'Text'
-- (@'RText' l@) here because information is presented in human-readable
-- form, i.e. percent-decoded, and thus it may contain Unicode characters.
data URI = URI
  { -- | URI scheme, if 'Nothing', then the URI reference is relative
    uriScheme :: Maybe (RText 'Scheme),
    -- | 'Authority' component in 'Right' or a 'Bool' value in 'Left'
    -- indicating if 'uriPath' path is absolute ('True') or relative
    -- ('False'); if we have an 'Authority' component, then the path is
    -- necessarily absolute, see 'isPathAbsolute'
    --
    -- __Note__: before version /0.1.0.0/ type of 'uriAuthority' was
    -- @'Maybe' 'Authority'@
    uriAuthority :: Either Bool Authority,
    -- | 'Nothing' represents the empty path, while 'Just' contains an
    -- indication 'Bool' whether the path component has a trailing slash,
    -- and the collection of path pieces @'NonEmpty' ('RText' 'PathPiece')@.
    --
    -- __Note__: before version /0.2.0.0/ type of 'uriPath' was @['RText'
    -- 'PathPiece']@.
    uriPath :: Maybe (Bool, NonEmpty (RText 'PathPiece)),
    -- | Query parameters, RFC 3986 does not define the inner organization
    -- of query string, so we deconstruct it following RFC 1866 here
    uriQuery :: [QueryParam],
    -- | Fragment, without @#@
    uriFragment :: Maybe (RText 'Fragment)
  }
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

-- | @since 0.3.5.0
instance Hashable URI

instance Arbitrary URI where
  arbitrary =
    URI
      <$> arbitrary
      <*> arbitrary
      <*> ( do
              mpieces <- NE.nonEmpty <$> arbitrary
              trailingSlash <- arbitrary
              return ((trailingSlash,) <$> mpieces)
          )
      <*> arbitrary
      <*> arbitrary

instance NFData URI

-- | @since 0.3.1.0
instance TH.Lift URI where
  lift = liftData
  liftTyped = TH.Code . TH.unsafeTExpCoerce . TH.lift

-- | Make a given 'URI' reference absolute using the supplied @'RText'
-- 'Scheme'@ if necessary.
makeAbsolute :: RText 'Scheme -> URI -> URI
makeAbsolute scheme URI {..} =
  URI
    { uriScheme = pure (fromMaybe scheme uriScheme),
      ..
    }

-- | Return 'True' if path in a given 'URI' is absolute.
--
-- @since 0.1.0.0
isPathAbsolute :: URI -> Bool
isPathAbsolute = fromLeft True . uriAuthority

-- | Authority component of 'URI'.
data Authority = Authority
  { -- | User information
    authUserInfo :: Maybe UserInfo,
    -- | Host
    authHost :: RText 'Host,
    -- | Port number
    authPort :: Maybe Word
  }
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

-- | @since 0.3.5.0
instance Hashable Authority

instance Arbitrary Authority where
  arbitrary =
    Authority
      <$> arbitrary
      <*> arbitrary
      <*> arbitrary

instance NFData Authority

-- | @since 0.3.1.0
instance TH.Lift Authority where
  lift = liftData
  liftTyped = TH.Code . TH.unsafeTExpCoerce . TH.lift

-- | User info as a combination of username and password.
data UserInfo = UserInfo
  { -- | Username
    uiUsername :: RText 'Username,
    -- | Password, 'Nothing' means that there was no @:@ character in the
    -- user info string
    uiPassword :: Maybe (RText 'Password)
  }
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

-- | @since 0.3.5.0
instance Hashable UserInfo

instance Arbitrary UserInfo where
  arbitrary =
    UserInfo
      <$> arbitrary
      <*> arbitrary

instance NFData UserInfo

-- | @since 0.3.1.0
instance TH.Lift UserInfo where
  lift = liftData
  liftTyped = TH.Code . TH.unsafeTExpCoerce . TH.lift

-- | Query parameter either in the form of flag or as a pair of key and
-- value. A key cannot be empty, while a value can.
data QueryParam
  = -- | Flag parameter
    QueryFlag (RText 'QueryKey)
  | -- | Key–value pair
    QueryParam (RText 'QueryKey) (RText 'QueryValue)
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

-- | @since 0.3.5.0
instance Hashable QueryParam

instance Arbitrary QueryParam where
  arbitrary =
    oneof
      [ QueryFlag <$> arbitrary,
        QueryParam <$> arbitrary <*> arbitrary
      ]

instance NFData QueryParam

-- | @since 0.3.1.0
instance TH.Lift QueryParam where
  lift = liftData
  liftTyped = TH.Code . TH.unsafeTExpCoerce . TH.lift

-- | Parse exception thrown by 'mkURI' when a given 'Text' value cannot be
-- parsed as a 'URI'.
newtype ParseException
  = -- | Arguments are: original input and parse error
    ParseException (ParseErrorBundle Text Void)
  deriving (Show, Eq, Data, Typeable, Generic)

instance Exception ParseException where
  displayException (ParseException b) = errorBundlePretty b

instance NFData ParseException

-- | Parse exception thrown by 'mkURIBs' when a given 'ByteString' value cannot be
-- parsed as a 'URI'.
--
-- @since 0.3.3.0
newtype ParseExceptionBs
  = -- | Arguments are: original input and parse error
    ParseExceptionBs (ParseErrorBundle ByteString Void)
  deriving (Show, Eq, Data, Typeable, Generic)

instance Exception ParseExceptionBs where
  displayException (ParseExceptionBs b) = errorBundlePretty b

instance NFData ParseExceptionBs

----------------------------------------------------------------------------
-- Refined text

-- | Refined text labelled at the type level.
newtype RText (l :: RTextLabel) = RText Text
  deriving (Eq, Ord, Data, Typeable, Generic)

-- | @since 0.3.5.0
instance Hashable (RText l)

instance Show (RText l) where
  show (RText txt) = show txt

instance NFData (RText l)

-- | @since 0.3.1.0
instance (Typeable l) => TH.Lift (RText l) where
  lift = liftData
  liftTyped = TH.Code . TH.unsafeTExpCoerce . TH.lift

-- | Refined text labels.
data RTextLabel
  = -- | See 'mkScheme'
    Scheme
  | -- | See 'mkHost'
    Host
  | -- | See 'mkUsername'
    Username
  | -- | See 'mkPassword'
    Password
  | -- | See 'mkPathPiece'
    PathPiece
  | -- | See 'mkQueryKey'
    QueryKey
  | -- | See 'mkQueryValue'
    QueryValue
  | -- | See 'mkFragment'
    Fragment
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

-- | This type class associates checking, normalization, and a term level
-- label with a label on the type level.
--
-- We would like to have a closed type class here, and so we achieve almost
-- that by not exporting 'RLabel' and 'mkRText' (only specialized helpers
-- like 'mkScheme').
class RLabel (l :: RTextLabel) where
  rcheck :: Proxy l -> Text -> Bool
  rnormalize :: Proxy l -> Text -> Text
  rlabel :: Proxy l -> RTextLabel

-- | Construct a refined text value.
mkRText :: forall m l. (MonadThrow m, RLabel l) => Text -> m (RText l)
mkRText txt =
  if rcheck lproxy txt
    then return . RText $ rnormalize lproxy txt
    else throwM (RTextException (rlabel lproxy) txt)
  where
    lproxy = Proxy :: Proxy l

-- | Lift a 'Text' value into @'RText' 'Scheme'@.
--
-- Scheme names consist of a sequence of characters beginning with a letter
-- and followed by any combination of letters, digits, plus @\"+\"@, period
-- @\".\"@, or hyphen @\"-\"@.
--
-- This smart constructor performs normalization of valid schemes by
-- converting them to lower case.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.1>
mkScheme :: (MonadThrow m) => Text -> m (RText 'Scheme)
mkScheme = mkRText

instance RLabel 'Scheme where
  rcheck Proxy = ifMatches $ do
    void . satisfy $ \x ->
      isAscii x && isAlpha x
    skipMany . satisfy $ \x ->
      isAscii x && isAlphaNum x || x == '+' || x == '-' || x == '.'
  rnormalize Proxy = T.toLower
  rlabel Proxy = Scheme

instance Arbitrary (RText 'Scheme) where
  arbitrary = arbScheme

-- | Lift a 'Text' value into @'RText' 'Host'@.
--
-- The host sub-component of authority is identified by an IP literal
-- encapsulated within square brackets, an IPv4 address in dotted-decimal
-- form, or a registered name.
--
-- This smart constructor performs normalization of valid hosts by
-- converting them to lower case.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.2.2>
mkHost :: (MonadThrow m) => Text -> m (RText 'Host)
mkHost = mkRText

instance RLabel 'Host where
  rcheck Proxy = (ifMatches . void . pHost) False
  rnormalize Proxy = T.toLower
  rlabel Proxy = Host

instance Arbitrary (RText 'Host) where
  arbitrary = arbHost

-- | Lift a 'Text' value into @'RText' 'Username'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.2.1>
mkUsername :: (MonadThrow m) => Text -> m (RText 'Username)
mkUsername = mkRText

instance RLabel 'Username where
  rcheck Proxy = not . T.null
  rnormalize Proxy = id
  rlabel Proxy = Username

instance Arbitrary (RText 'Username) where
  arbitrary = arbText' mkUsername

-- | Lift a 'Text' value into @'RText' 'Password'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.2.1>
mkPassword :: (MonadThrow m) => Text -> m (RText 'Password)
mkPassword = mkRText

instance RLabel 'Password where
  rcheck Proxy = const True
  rnormalize Proxy = id
  rlabel Proxy = Password

instance Arbitrary (RText 'Password) where
  arbitrary = arbText mkPassword

-- | Lift a 'Text' value into @'RText' 'PathPiece'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.3>
mkPathPiece :: (MonadThrow m) => Text -> m (RText 'PathPiece)
mkPathPiece = mkRText

instance RLabel 'PathPiece where
  rcheck Proxy = not . T.null
  rnormalize Proxy = id
  rlabel Proxy = PathPiece

instance Arbitrary (RText 'PathPiece) where
  arbitrary = arbText' mkPathPiece

-- | Lift a 'Text' value into @'RText 'QueryKey'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.4>
mkQueryKey :: (MonadThrow m) => Text -> m (RText 'QueryKey)
mkQueryKey = mkRText

instance RLabel 'QueryKey where
  rcheck Proxy = not . T.null
  rnormalize Proxy = id
  rlabel Proxy = QueryKey

instance Arbitrary (RText 'QueryKey) where
  arbitrary = arbText' mkQueryKey

-- | Lift a 'Text' value into @'RText' 'QueryValue'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.4>
mkQueryValue :: (MonadThrow m) => Text -> m (RText 'QueryValue)
mkQueryValue = mkRText

instance RLabel 'QueryValue where
  rcheck Proxy = const True
  rnormalize Proxy = id
  rlabel Proxy = QueryValue

instance Arbitrary (RText 'QueryValue) where
  arbitrary = arbText mkQueryValue

-- | Lift a 'Text' value into @'RText' 'Fragment'@.
--
-- This smart constructor does not perform any sort of normalization.
--
-- See also: <https://tools.ietf.org/html/rfc3986#section-3.5>
mkFragment :: (MonadThrow m) => Text -> m (RText 'Fragment)
mkFragment = mkRText

instance RLabel 'Fragment where
  rcheck Proxy = const True
  rnormalize Proxy = id
  rlabel Proxy = Fragment

instance Arbitrary (RText 'Fragment) where
  arbitrary = arbText mkFragment

-- | Project a plain strict 'Text' value from a refined @'RText' l@ value.
unRText :: RText l -> Text
unRText (RText txt) = txt

-- | The exception is thrown when a refined @'RText' l@ value cannot be
-- constructed due to the fact that given 'Text' value is not correct.
data RTextException
  = -- | 'RTextLabel' identifying what sort of refined text value could not be
    -- constructed and the input that was supplied, as a 'Text' value
    RTextException RTextLabel Text
  deriving (Show, Eq, Ord, Data, Typeable, Generic)

instance Exception RTextException where
  displayException (RTextException lbl txt) =
    "The value \""
      ++ T.unpack txt
      ++ "\" could not be lifted into a "
      ++ show lbl

----------------------------------------------------------------------------
-- Parser helpers

-- | Return 'True' if given parser can consume 'Text' in its entirety.
ifMatches :: Parsec Void Text () -> Text -> Bool
ifMatches p = isJust . parseMaybe p

----------------------------------------------------------------------------
-- Arbitrary helpers

-- | Generator of 'Arbitrary' schemes.
arbScheme :: Gen (RText 'Scheme)
arbScheme = do
  let g = oneof [choose ('a', 'z'), choose ('A', 'Z')]
  x <- g
  xs <-
    listOf $
      frequency [(3, g), (1, choose ('0', '9'))]
  return . fromJust . mkScheme . T.pack $ x : xs

-- | Generator of 'Arbitrary' hosts.
arbHost :: Gen (RText 'Host)
arbHost =
  fromJust . mkHost . T.pack
    <$> frequency
      [ (1, ipLiteral),
        (2, ipv4Address),
        (4, regName),
        (1, return "")
      ]
  where
    ipLiteral = do
      xs <- oneof [ipv6Address, ipvFuture]
      return ("[" ++ xs ++ "]")
    ipv6Address =
      -- NOTE We do not mess with zeroes here, because it's a hairy stuff.
      -- We test how we handle :: thing manually in the test suite.
      intercalate ":" . fmap (`showHex` "")
        <$> vectorOf 8 (arbitrary :: Gen Word16)
    ipv4Address =
      intercalate "." . fmap (`showInt` "")
        <$> vectorOf 4 (arbitrary :: Gen Word8)
    ipvFuture = do
      v <- oneof [choose ('0', '9'), choose ('a', 'f')]
      xs <-
        listOf1 $
          frequency
            [ (3, choose ('a', 'z')),
              (3, choose ('A', 'Z')),
              (2, choose ('0', '9')),
              (2, elements "-._~!$&'()*+,;=:")
            ]
      return ("v" ++ [v] ++ "." ++ xs)
    domainLabel = do
      let g = arbitrary `suchThat` isUnreservedChar
      x <- g
      xs <-
        listOf $
          frequency [(3, g), (1, return '-')]
      x' <- g
      return ([x] ++ xs ++ [x'])
    regName = intercalate "." <$> resize 5 (listOf1 domainLabel)

-- | Return 'True' if the given character is unreserved.
isUnreservedChar :: Char -> Bool
isUnreservedChar x =
  isAlphaNum x || x == '-' || x == '.' || x == '_' || x == '~'

-- | Make generator for refined text given how to lift a possibly empty
-- arbitrary 'Text' value into a refined type.
arbText :: (Text -> Maybe (RText l)) -> Gen (RText l)
arbText f = fromJust . f . T.pack <$> listOf arbitrary

-- | Like 'arbText'', but the lifting function will be given non-empty
-- arbitrary 'Text' value.
arbText' :: (Text -> Maybe (RText l)) -> Gen (RText l)
arbText' f = fromJust . f . T.pack <$> listOf1 arbitrary

----------------------------------------------------------------------------
-- TH lifting helpers

liftData :: (Data a, TH.Quote m) => a -> m TH.Exp
liftData = TH.dataToExpQ (fmap liftText . cast)

liftText :: (TH.Quote m) => Text -> m TH.Exp
liftText t = TH.AppE (TH.VarE 'T.pack) <$> TH.lift (T.unpack t)
